src/ir/term/mod.rs

```diff
@@ -1252,175 +1258,227 @@ impl Value {
+            let mut term_vecs = vec![Vec::new(); arr_size];
+            //2D vector: term_vecs[i] will store a vector of all the i-th index
+            //  entries of the array arguments
+
+            //Value::BitVector(BitVector::new(uint.into(),width,))
+            for i in 0..arg_cnt {
+                let arr = vs.get(&c.cs[i]).unwrap().as_array().clone();
+                for j in 0..arr_size {
+                    let jval = &Value::BitVector(BitVector::new(Integer::from(j), 32));
+                    let term = leaf_term(Op::Const(arr.clone().select(jval)));
+                    term_vecs[j].push(term);
                 }
-            }),
-            Op::PfNaryOp(o) => Value::Field({
-                let mut xs = c.cs.iter().map(|c| vs.get(c).unwrap().as_pf().clone());
-                let f = xs.next().unwrap();
-                xs.fold(
-                    f,
-                    match o {
-                        PfNaryOp::Add => std::ops::Add::add,
-                        PfNaryOp::Mul => std::ops::Mul::mul,
-                    },
-                )
-            }),
-            Op::UbvToPf(m) => Value::Field({
-                let a = vs.get(&c.cs[0]).unwrap().as_bv().clone();
-                field::FieldElem::new(a.uint().clone(), m.clone())
-            }),
-            // tuple
-            Op::Tuple => Value::Tuple(c.cs.iter().map(|c| vs.get(c).unwrap().clone()).collect()),
-            Op::Field(i) => {
-                let t = vs.get(&c.cs[0]).unwrap().as_tuple();
-                assert!(i < &t.len(), "{} out of bounds for {}", i, c.cs[0]);
-                t[*i].clone()
-            }
-            Op::Update(i) => {
-                let mut t = Vec::from(vs.get(&c.cs[0]).unwrap().as_tuple()).into_boxed_slice();
-                assert!(i < &t.len(), "{} out of bounds for {}", i, c.cs[0]);
-                let e = vs.get(&c.cs[1]).unwrap().clone();
-                assert_eq!(t[*i].sort(), e.sort());
-                t[*i] = e;
-                Value::Tuple(t)
-            }
-            // array
-            Op::Store => {
-                let a = vs.get(&c.cs[0]).unwrap().as_array().clone();
-                let i = vs.get(&c.cs[1]).unwrap().clone();
-                let v = vs.get(&c.cs[2]).unwrap().clone();
-                Value::Array(a.clone().store(i, v))
             }
-            Op::Select => {
-                let a = vs.get(&c.cs[0]).unwrap().as_array().clone();
-                let i = vs.get(&c.cs[1]).unwrap();
-                a.clone().select(i)
+            //Cloning the first arg just for formatting
+            let mut res = vs.get(&c.cs[0]).unwrap().as_array().clone();
+            for i in 0..arr_size {
+                let t = term((**op).clone(), term_vecs[i].clone());
+                let val = eval_value(vs, h, t);
+                res.map
+                    .insert(Value::BitVector(BitVector::new(Integer::from(i), 32)), val);
```

Comment:
I'd like to avoid assuming that the array is indexed by `(bv 32)`. I think we should accept
any BV and prime-field element, and any boolean.

Reply:
When the indices are bitvectors, it is simple to iterate through the terms, as I can use a for loop.
How should I iterate through the indices for these other index types?

Reply:
You can do a similar thing for boolean and field elements: https://github.com/circify/circ/blob/master/src/ir/term/mod.rs#L835-L838

Reply:
Just to clarify, should I case on the index type and write separate iterating code for each?

Reply:
I was going to say that you can just call `elems_iter`, but I now see that doesn't *quite* work.

You should extract a value-based (rather than term-based) version of `elems_iter`, and call the extracted function.

Reply:
Got it.
When I look at elems_iter, it looks as though it iterates through all of the possible values of the index type.

I am assuming that in my iterator for map, the number of elements I iterate should depend on the size of the arrays I am mapping on. In the case where the key_type is a bitvector or field and the array size is x, can I assume that the indices are going to be 0,1,2,...x-1?

Reply:
Yep!